libflux/src/semantic/types.rs

```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+            (t, MonoType::Var(tv)) => tv.unify(t, cons),
+            (MonoType::Arr(t), MonoType::Arr(s)) => t.unify(*s, cons),
+            (MonoType::Row(t), MonoType::Row(s)) => t.unify(*s, cons),
+            (MonoType::Fun(t), MonoType::Fun(s)) => t.unify(*s, cons),
+            (t, with) => Err(Error::Unify { t, with }),
+        }
+    }
+
+    fn constrain(self, with: HashSet<Kind>, cons: Const) -> Result {
+        match self {
+            MonoType::Bool => self
+                .satisfies_at_most(maplit::hashset![Kind::Equatable, Kind::Nullable,])(
+                with, cons
+            ),
+            MonoType::Int => self.satisfies_at_most(maplit::hashset![
+                Kind::Addable,
+                Kind::Subtractable,
+                Kind::Divisible,
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
+            MonoType::Uint => self.satisfies_at_most(maplit::hashset![
+                Kind::Addable,
+                Kind::Divisible,
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
+            MonoType::Float => self.satisfies_at_most(maplit::hashset![
+                Kind::Addable,
+                Kind::Subtractable,
+                Kind::Divisible,
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
+            MonoType::String => self.satisfies_at_most(maplit::hashset![
+                Kind::Addable,
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
+            MonoType::Duration => self.satisfies_at_most(maplit::hashset![
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable,
+            ])(with, cons),
+            MonoType::Time => self.satisfies_at_most(maplit::hashset![
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable,
+            ])(with, cons),
+            MonoType::Regexp => Err(Error::Constrain { t: self, with }),
+            MonoType::Var(tvr) => tvr.constrain(with, cons),
+            MonoType::Arr(arr) => arr.constrain(with, cons),
+            MonoType::Row(obj) => obj.constrain(with, cons),
+            MonoType::Fun(fun) => fun.constrain(with, cons),
+        }
+    }
+
+    // Returns a closure that specifies the possible type classes (kinds) to which a monotype belongs
+    fn satisfies_at_most(
+        self,
+        kinds: HashSet<Kind>,
+    ) -> impl FnOnce(HashSet<Kind>, Const) -> Result {
+        move |k, c| {
+            if k.is_subset(&kinds) {
+                Ok((Subst::empty(), c))
+            } else {
+                Err(Error::Constrain {
+                    t: self,
+                    with: k.difference(&kinds).map(|&kind| kind).collect(),
```

Comment:
what does the `map` call do here? I looks like the identity as far as I can tell.

Reply:
Basically. The error enum owns the kinds passed into it, so I have to collect a set of kinds vs a set of references to kinds. The error enum could probably use a little more work.

Reply:
ok so it converts `[Kind]` to `[&Kind]`? Should we instead use `&[Kind]`?

Reply:
It converts `[&Kind]` to `[Kind]`. I think it makes sense for errors to consume their kinds. The pattern is that kinds are owned, they're not referenced, so the same should hold for errors.